File: zZ/cpp/binary-tree-postorder-traversal.cpp
// 145. Binary Tree Postorder Traversal
// Difficulty: Hard

// Given a binary tree, return the postorder traversal of its nodes values.

// For example:
// Given binary tree {1,#,2,3},
//    1
//     \
//      2
//     /
//    3
// return [3,2,1].

// Note: Recursive solution is trivial, could you do it iteratively?

// Time:  O(n)
// Space: O(h)
class Solution {
public:
  vector<int> postorderTraversal(TreeNode* root) {
    vector<int> result;
    stack<const TreeNode*> s;
    const TreeNode* curr = root;
    do {
      while (curr) {
        s.push(curr);
        curr = curr->left;
      }
      TreeNode* prev = nullptr;
      while (!s.empty()) {
        curr = s.top(); s.pop();

        if (curr->right == prev) {
          result.push_back(curr->val);
          prev = curr;
        } else {
          s.push(curr);
          curr = curr->right;
          break;
        }
      }
    } while (!s.empty());
    return result;
  }
};

File: zZ/java/binary-tree-postorder-traversal.java
/**
 * Binary Tree Postorder Traversal
 * 
 * Given a binary tree, return the postorder traversal of its nodes' values.
 * 
 * For example:
 * Given binary tree {1,#,2,3},
 *    1
 *     \
 *      2
 *     /
 *    3
 * return [3,2,1].
 * 
 * Note: Recursive solution is trivial, could you do it iteratively?
 */

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

public class Solution {
    public List<Integer> postorderTraversal(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        Stack<TreeNode> stack = new Stack<>();
        TreeNode curr = root;

        do {
            while (curr.left != null) {
                stack.push(curr);
                curr = curr.left;
            }

            TreeNode prev = null;
            while (!stack.isEmpty()) {
                curr = stack.pop();

                if (prev == curr.right) {
                    result.add(curr.val);
                    prev = curr;
                } else {
                    stack.push(curr);
                    curr = curr.right;
                    break;
                }

            }
        } while (!stack.isEmpty());

        return result;
    }
}

public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }